include/deal.II/matrix_free/matrix_free.h

```diff
@@ -3321,13 +3320,15 @@ namespace internal
 
 
     /**
-     * Zero out vector region for vector that do _not_ support
-     * exchange on a subset of DoFs <==> begin() + ind == local_element(ind)
+     * Zero out vector region for vector that do _not_ support exchange on a
+     * subset of DoFs <==> begin() + ind == local_element(ind) but are still a
+     * vector type
      */
     template <
       typename VectorType,
       typename std::enable_if<!has_exchange_on_subset<VectorType>::value,
-                              VectorType>::type * = nullptr>
+                              VectorType>::type * = nullptr,
+      typename VectorType::value_type * = nullptr>
```

Comment:
> The matrix-free loops previously also supported the case where the VectorType that is passed in as a second/third argument to cell_loop() and loop() is not actually a vector but say a dummy.

In other places I catch it with `is_serial_or_dummy` type-trait, but that's not exactly what we need here, I guess. What we need just `is_dummy`.

May I suggest an alternative solution, namely add more type-trait `is_dummy` which is based on existence of `T::value_type` and use it here as `&& !is_dummy<VectorType>::value` and below you would have `enable_if<is_dummy<>::value>` variant that shall never be called. 

Otherwise the current code will still fail if someone tries to feed `int` instead of `unsigned int`.

Reply:
we would then have
```
template <
	      typename VectorType,
	      typename std::enable_if<!has_exchange_on_subset<VectorType>::value && !is_dummy<VectorType>::value,
	                              VectorType>::type * = nullptr>
```
here.

Reply:
> Otherwise the current code will still fail if someone tries to feed `int` instead of `unsigned int`.

How do you mean? I use `...` as the argument, the `unsigned int` in the default block is for the range number (that is not used in the dummy function).

I am not opposed to changing it, but to me the current solution appears more natural, as the alternative would mean to introduce yet another `struct` in the type traits class when all that we need is already there - we can catch for `...::value_type` in the template argument list to trigger SFINAE.

Reply:
> How do you mean? I use ... as the argument and it should work for all argument types that do not match one of the two other functions, the unsigned int in the default block is for the range number (that is not used in the dummy function).

oh, I confused this type with `VectorType`, you are right.

> but to me the current solution appears more natural,

ok. I don't have a strong opinion on this. For me the reason for extra type-trait was motivated by having everything in once place. In an unlikely event if eventually we have vectors without `::value_type`, we know what to change in one single place to keep things working.

Reply:
I would leave things simple right now - in my opinion, the reason why we have the third option is actually clearer at the spot we have it since we have the `typename VectorType::value_type()` just a few lines above. But I agree that we should centralize these decisions once they get spread out more. Let us both keep an eye on it.

Reply:
sounds good to me.